src/storage/src/source/postgres.rs

```diff
@@ -128,28 +126,292 @@ macro_rules! try_recoverable {
     };
 }
 
-impl PostgresSourceReader {
-    /// Constructs a new instance
-    pub fn new(
+// Message used to communicate between `get_next_message` and the tokio task
+enum InternalMessage {
+    Err(SourceReaderError),
+    Value {
+        value: Row,
+        lsn: PgLsn,
+        diff: Diff,
+        end: bool,
+    },
+}
+
+/// Information required to sync data from Postgres
+pub struct PostgresSourceReader {
+    receiver_stream: Receiver<InternalMessage>,
+}
+
+/// An internal struct held by the spawned tokio task
+struct PostgresTaskInfo {
+    source_id: GlobalId,
+    connector: PostgresSourceConnector,
+    /// Our cursor into the WAL
+    lsn: PgLsn,
+    metrics: PgSourceMetrics,
+    source_tables: HashMap<u32, PostgresTable>,
+    sender: Sender<InternalMessage>,
+    activator: SyncActivator,
+}
+
+impl SourceReader for PostgresSourceReader {
+    type Key = ();
+    type Value = Row;
+    // Postgres can produce deletes that cause retractions
+    type Diff = Diff;
+
+    fn new(
+        _source_name: String,
         source_id: GlobalId,
-        connector: PostgresSourceConnector,
-        metrics: &SourceBaseMetrics,
-    ) -> Self {
-        Self {
-            source_id,
+        _worker_id: usize,
+        _worker_count: usize,
+        consumer_activator: SyncActivator,
+        connector: ExternalSourceConnector,
+        _restored_offsets: Vec<(PartitionId, Option<MzOffset>)>,
+        _encoding: SourceDataEncoding,
+        metrics: SourceBaseMetrics,
+        _connector_context: ConnectorContext,
+    ) -> Result<Self, anyhow::Error> {
+        let connector = match connector {
+            ExternalSourceConnector::Postgres(pg) => pg,
+            _ => {
+                panic!("Postgres is the only legitimate ExternalSourceConnector for PostgresSourceReader")
+            }
+        };
+
+        let (dataflow_tx, dataflow_rx) = tokio::sync::mpsc::channel(10_000);
```

Comment:
yes, it will just put the pg task to sleep till it has space (but a later pr will probably remove the task and move it to the `SourceReader::next` api